mesonbuild/interpreter.py

```diff
@@ -3171,7 +3171,10 @@ def func_project(self, node, args, kwargs):
         # have any effect.
         self.project_default_options = mesonlib.stringlistify(kwargs.get('default_options', []))
         self.project_default_options = coredata.create_options_dict(self.project_default_options, self.subproject)
-        if self.environment.first_invocation:
+
+        # For the main project we use first_invocation to know whether to init builtins,
+        # for subprojects see if the surpbojrect is in the known subprojects
+        if (self.subproject == '' and not self.environment.first_invocation) or self.subproject not in self.subprojects:
```

Comment:
I don't think you got that condition right. `self.subproject not in self.subprojects` that's always going to be true when subproject is `''`

Reply:
Gah, yeah. It's annoying how many cases of "we need the same information about supprojects and the main project, but we encode them differently